.editorconfig

```diff
@@ -424,10 +424,10 @@ dotnet_diagnostic.SA1508.severity = error
 dotnet_diagnostic.SA1509.severity = error
 dotnet_diagnostic.SA1510.severity = error
 dotnet_diagnostic.SA1511.severity = error
-dotnet_diagnostic.SA1512.severity = error
```

Comment:
Let's see...

RCS1085 is for always using auto-implemented properties, when possible. Violations are [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs#L92), [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForStream.cs#L87), and [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/List/Internal/ExpireAfter.cs#L87). Here we have a complex object where there's far more value in being able to see all of its persistent state (I.E. private fields) in one place, than in reducing the file size by 2 lines. This is a common-enough scenario, in my mind, that it warrants disabling the rule.

SA1201 says "a property should not follow a method", but it ignores the fact that it's a `protected` property, and I've always encountered standard practice to be that `protected` instance members don't come before `public` instance members. Violations are [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs#L92), [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForStream.cs#L87), and [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/List/Internal/ExpireAfter.cs#L87).

SA1512 says "single-line comments should not be followed by blank line", with violations [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs#L316), [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForStream.cs#L239), [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForStream.cs#L317), and [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/List/Internal/ExpireAfter.cs#L320). Intuitively, it's standard practice to "attach" comments to the lines of code that they apply to, which I imagine is the origin of this rule, but all of these are examples of comments that DON'T apply to a specific line of code. They are actually commenting on the LACK of lines of code that one might expect to be here.

SA1513 says "Closing brace should be followed by blank line", and is violated [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs#L162) and [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForStream.cs#L235). In these cases, I'm attaching a line of code to the end of a loop, in an effort to indicate that it fundamentally is a finalization action for that loop. Not a big deal, I probably would have just fixed these two if I hadn't already disabled the rule for the 11 other violations (I'll just link [this one](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs#L247), they're all the same), which consist of having a `break;` immediately following a bracketed `case` block. In my mind the closing bracket itself serves the purpose that a blank line normally would, for a multi-line `case` block, and inserting an additional blank line here makes it subtly harder to read.

SA1515 says "Single line comments should be preceeded by blank line", and is violated [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs#L73), [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForSource.cs#L109), [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/Cache/Internal/ExpireAfter.ForStream.cs#L104), and [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData/List/Internal/ExpireAfter.cs#L68). Also [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData.Tests/Utilities/FakeScheduler.cs#L58) and [here](https://github.com/reactivemarbles/DynamicData/blob/f73153d511adad4e285239d0b7e7d5027e77652e/src/DynamicData.Tests/Utilities/FakeScheduler.cs#L64), but those are within the `.Tests` project. In all of these cases, we're commenting on a single line withing a multi-line action, where the comment is too long to put at the end of the line, and where the multi-line action would normally have no blank lines in it. These are all valid in my mind, because of the implicit rule that comments should be directly attached to the thing they're commenting on, not potentially many lines away from it, and with so many valid use-cases being flagged, this seems like a rule that should be disabled.

Reply:
I got to be honest, I really dislike style enforcement. 

I have been tempted to turn it off entirely in DD as it drives me insane how slow it is to run and how it tries to force me to comment code which I have not even written yet!!!!  I always end up turning it off, do my code, then turn it on to see what it complains about.  The worst thing is that it also slows my development down.

We're grown ups and surely we should be allowed to express out own style.  There are a few things that are important like comments, but surely we could enforce that as part of the PR process.

In a better version of of the net eco-system we'd all be using automatic linting. 

Perhaps we should get rid of it completely  - thoughts?

Reply:
> I got to be honest, I really dislike style enforcement.

I fully agree. I think style-enforcement is kind of a fundamentally-flawed concept. Good styling is about promoting human-readability of code, computers are fundamentally ill-equipped to understand what is "human readable", across the board. At least, I have yet to see any system that implements rules granular enough to promote readability, instead of harm it.

> Perhaps we should get rid of it completely - thoughts?

I wouldn't lose any sleep, personally. But on a large collaborative project, logically, I know there's a bigger picture than just my opinions, and it would probably be a mistake to remove completely. I think we can focus on de-tuning it, and reducing the impact it has on the developer experience.

> I always end up turning it off, do my code, then turn it on to see what it complains about.

What's the quick-and-easy way to do this? I've just been manually tossing out `#pragma warning disable` as they pop up, and it's quite annoying.

> we'd all be using automatic linting.

So, wait, what is linting if not this? Like, automatically fixing violations, instead of just reporting them?